### PR TITLE
[MRG+1-1] Check masker utils

### PR DESCRIPTION
**Pull request by ahoyosid**
We refactor the screening_percentile parameter, putting it in _utils. 
Additionally, we  put create a validation file where the check masking utils is contained.
The check_masking  is used in two PRs: #1336 and #698

Can someone review this ?
@KamalakerDadi @GaelVaroquaux @AlexandreAbraham @mrahim

### CODE REVIEW

File: nilearn/_utils/param_validation.py [hunk 1]
```diff
@@ -63,3 +68,75 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
...
+    ----------
+    mask_img : nibabel image object
+        Input image whose voxel dimensions are to be computed.
+
+    Returns
+    -------
+    vol : float
+        The computed volume.
+    """
+    vox_dims = mask_img.get_header().get_zooms()[:3]
```
Comment (GaelVaroquaux) [inline nilearn/_utils/param_validation.py]: You need to use get_header from nilearn._utils.compat here, to avoid problems with multiple versions of nibabel.

Comment (ahoyosid) [inline nilearn/_utils/param_validation.py]: Ok

File: nilearn/_utils/param_validation.py [hunk 2]
```diff
@@ -63,3 +68,75 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
         raise TypeError('%s should be either a number '
                         'or a string finishing with a percent sign' % (name, ))
     return threshold
+
+
+def _get_mask_volume(mask_img):
+    """Computes the volume of a brain mask in mm^3
```
Comment (GaelVaroquaux) [inline nilearn/_utils/param_validation.py]: Missing empty line here.

File: nilearn/_utils/param_validation.py [hunk 3]
```diff
@@ -63,3 +68,77 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
...
+
+    Retruns
+    -------
+    screening_percentile: float in the interval [0, 100]
+        Percentile value for ANOVa univariate feature selection.
+    """
+    original_screening_percentile = screening_percentile
+    # correct screening_percentile according to the volume of the data mask
+    mask_volume = _get_mask_volume(mask_img)
+    if mask_volume > MNI152_BRAIN_VOLUME:
```
Comment (bthirion) [inline nilearn/_utils/param_validation.py]: I find this threshold a bit strange. This has 50% chance to occur in real life... ?
I would test if mask_volme < 1.1 * MNI152_BRAIN_VOLUME
and Mask_volume >.5 * MNI152_BRAIN_VOLUME
Maybe I'm missing something.

Comment (GaelVaroquaux) [inline nilearn/_utils/param_validation.py]: That's a good point. I suggest changing this line to "1.1 * MNI152_BRAIN_VOLUME"

Comment (ahoyosid) [inline nilearn/_utils/param_validation.py]: I think checking a coverage greater than 0.5 MNI152_BRAIN_VOLUME is a good sanity check. I mean, someone can use the mask of a ROI.

**Comment by GaelVaroquaux** [comment]
> I think checking a coverage greater than 0.5 MNI152_BRAIN_VOLUME is a good
> sanity check.

You mean: you don't think?

> I mean, someone can use the mask of a ROI.

Indeed, that's why we kept it that way.

**Comment by ahoyosid** [comment]
Ok

**Comment by bthirion** [comment]
OK, but since this is only a warning, I thought that this might be a 
hint that the user is using a functiona that was intended to be used in 
a full brain setting.
On the other hand, spitting too many warnings may be bad too. I don't 
have a strong opinion on this.

File: nilearn/_utils/param_validation.py [hunk 4]
```diff
@@ -63,3 +68,77 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
...
+    screening_percentile: float in the interval [0, 100]
+        Percentile value for ANOVa univariate feature selection.
+    """
+    original_screening_percentile = screening_percentile
+    # correct screening_percentile according to the volume of the data mask
+    mask_volume = _get_mask_volume(mask_img)
+    if mask_volume > 1.1 * MNI152_BRAIN_VOLUME:
+        warnings.warn(
+            "Brain mask is bigger than the volume of a standard "
+            "human brain. SpaceNet is probably not tuned to "
```
Comment (bthirion) [inline nilearn/_utils/param_validation.py]: Actually, this module should be independent from SpaceNet. Do you believe that the use will be confined to the future Decoder object, in which case SpaceNet -> Decoder ? Otherwise, we should remove the reference to SpaceNet.

**Comment by GaelVaroquaux** [comment]
> Otherwise, we should remove the reference to SpaceNet.

Yes. Replace it by "this object".

**Comment by GaelVaroquaux** [comment]
LGTM. +1 for merge. Thanks

File: nilearn/_utils/param_validation.py [hunk 5]
```diff
@@ -63,3 +68,77 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
...
+        mask of the data and volume of a standard brain.  If '100' is given,
+        all the features are used, regardless of the number of voxels.
+
+    mask_img : nibabel image object
+        Input image whose voxel dimensions are to be computed.
+
+    verbose : int, optional (default 0)
+        Verbosity level.
+
+    Retruns
```
Comment (AlexandreAbraham) [inline nilearn/_utils/param_validation.py]: Typo

File: nilearn/_utils/validation.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+from .param_validation import _adjust_screening_percentile
+from .compat import _basestring
+from ..input_data import NiftiMasker, MultiNiftiMasker
+
+
+def check_masking(mask, target_affine=None, target_shape=None,
+                  smoothing_fwhm=None, standardize=True,
+                  mask_strategy='epi', memory=None, memory_level=1):
+    """Setup a nifti masker.
```
Comment (AlexandreAbraham) [inline nilearn/_utils/validation.py]: Empty line

File: nilearn/_utils/validation.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    # mask is a masker object
+    elif isinstance(mask, (NiftiMasker, MultiNiftiMasker)):
+        try:
+            masker = clone(mask)
+            if hasattr(mask, 'mask_img_'):
+                mask_img = mask.mask_img_
+                masker.set_params(mask_img=mask_img)
+                masker.fit()
+        except TypeError as e:
+            # Workaround for a joblib bug: in joblib 0.6, a Memory object
```
Comment (AlexandreAbraham) [inline nilearn/_utils/validation.py]: But then, in this case, `mask.mask_img` is not updated... This should go above line 80 I believe.

File: nilearn/_utils/validation.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+from sklearn.externals.joblib import Memory
+from sklearn.feature_selection import (SelectPercentile, f_regression,
+                                       f_classif)
+
+from .param_validation import _adjust_screening_percentile
+from .compat import _basestring
+from ..input_data import NiftiMasker, MultiNiftiMasker
+
+
+def check_masking(mask, target_affine=None, target_shape=None,
```
Comment (AlexandreAbraham) [inline nilearn/_utils/validation.py]: Naming is inconsistent with `check_niimg` IMHO. This does not simply load the mask and do sanity checks.

File: nilearn/_utils/validation.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        means more memory for caching.
+
+    Returns
+    -------
+    masker : NiftiMasker or MultiNiftiMasker instance
+        Mask to be used on data.
+    """
+    # mask is an image, not a masker
+    if isinstance(mask, _basestring) or (mask is None):
+        masker = NiftiMasker(mask_img=mask,
```
Comment (AlexandreAbraham) [inline nilearn/_utils/validation.py]: In this case the masker is not fitted. Is this the intended behavior?

File: nilearn/decoding/tests/test_space_net.py [hunk 10]
```diff
@@ -283,16 +280,6 @@ def test_univariate_feature_screening(dim=(11, 12, 13), n_samples=10):
...
 
 
-def test_get_mask_volume():
-    # Test that hard-coded standard mask volume can be corrected computed
-    if os.path.isfile(mni152_brain_mask):
-        assert_equal(MNI152_BRAIN_VOLUME, _get_mask_volume(nibabel.load(
-            mni152_brain_mask)))
-    else:
-        warnings.warn("Couldn't find %s (for testing)" % (
-            mni152_brain_mask))
```
Comment (GaelVaroquaux) [inline nilearn/decoding/tests/test_space_net.py]: Why did you remove this? You should readd it somewhere else.

**Comment by GaelVaroquaux** [comment]
I have just realized that there is a bit of a problem with this PR: at first glance it duplicates the functionality in nilearn.input_data.masker_validation.check_embedded_nifti_masker

**Comment by ahoyosid** [comment]
Can you check this PR?

ping @AlexandreAbraham @GaelVaroquaux

File: nilearn/_utils/param_validation.py [hunk 11]
```diff
@@ -63,3 +68,77 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
...
+    mask_img : nibabel image object
+        Input image whose voxel dimensions are to be computed.
+
+    Returns
+    -------
+    vol : float
+        The computed volume.
+    """
+    vox_dims = get_header(mask_img).get_zooms()[:3]
+    return 1. * np.prod(vox_dims) * mask_img.get_data().astype(np.bool).sum()
```
Comment (bthirion) [inline nilearn/_utils/param_validation.py]: You're assuming an orthonormal grid, aren't you. Why don't you take
np.abs(np.linalg.det (make_img.affine[:3, :3])) instead of np.prod(vox_dims) ?

Comment (bthirion) [inline nilearn/_utils/param_validation.py]: LGTM otherwise

**Comment by KamalakerDadi** [comment]
You moved functions to nilearn/_utils/param_validation.py and testing file nilearn/tests/test_validation.py in _utils which doesn't exist and wrong path. 

Correct path to add tests is nilearn/tests/test_XXXXX.py

Are you sure you want to move functions to param_validation.py file or create a new file named as validation.py ?

File: nilearn/_utils/param_validation.py [hunk 12]
```diff
@@ -63,3 +68,77 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
...
+
+def _adjust_screening_percentile(screening_percentile, mask_img,
+                                 verbose=0):
+    """Adjusts the screening percentile according to the MNI mask template.
+
+    Parameters
+    ----------
+    screening_percentile : float in the interval [0, 100]
+        Percentile value for ANOVA univariate feature selection. A value of
+        100 means 'keep all features'. This percentile is is expressed
```
Comment (KamalakerDadi) [inline nilearn/_utils/param_validation.py]: is is -> is

File: nilearn/_utils/param_validation.py [hunk 13]
```diff
@@ -63,3 +68,77 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
...
+    vol : float
+        The computed volume.
+    """
+    vox_dims = get_header(mask_img).get_zooms()[:3]
+    return 1. * np.prod(vox_dims) * mask_img.get_data().astype(np.bool).sum()
+
+
+def _adjust_screening_percentile(screening_percentile, mask_img,
+                                 verbose=0):
+    """Adjusts the screening percentile according to the MNI mask template.
```
Comment (KamalakerDadi) [inline nilearn/_utils/param_validation.py]: May be MNI mask template -> MNI152 template.

File: nilearn/_utils/param_validation.py [hunk 14]
```diff
@@ -63,3 +68,77 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
...
+
+    Parameters
+    ----------
+    screening_percentile : float in the interval [0, 100]
+        Percentile value for ANOVA univariate feature selection. A value of
+        100 means 'keep all features'. This percentile is is expressed
+        w.r.t the volume of a standard (MNI152) brain, and so is corrected
+        at runtime by premultiplying it with the ratio of the volume of the
+        mask of the data and volume of a standard brain.  If '100' is given,
+        all the features are used, regardless of the number of voxels.
```
Comment (KamalakerDadi) [inline nilearn/_utils/param_validation.py]: This sentence is replicated. You can add remaining "regardless of the number of voxels." to start.

File: nilearn/_utils/param_validation.py [hunk 15]
```diff
@@ -63,3 +68,77 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
...
+    mask_img : nibabel image object
+        Input image whose voxel dimensions are to be computed.
+
+    verbose : int, optional (default 0)
+        Verbosity level.
+
+    Returns
+    -------
+    screening_percentile: float in the interval [0, 100]
+        Percentile value for ANOVa univariate feature selection.
```
Comment (KamalakerDadi) [inline nilearn/_utils/param_validation.py]: ANOVa -> ANOVA

File: nilearn/_utils/param_validation.py [hunk 16]
```diff
@@ -63,3 +68,77 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
...
+    """Adjusts the screening percentile according to the MNI mask template.
+
+    Parameters
+    ----------
+    screening_percentile : float in the interval [0, 100]
+        Percentile value for ANOVA univariate feature selection. A value of
+        100 means 'keep all features'. This percentile is is expressed
+        w.r.t the volume of a standard (MNI152) brain, and so is corrected
+        at runtime by premultiplying it with the ratio of the volume of the
+        mask of the data and volume of a standard brain.  If '100' is given,
```
Comment (KamalakerDadi) [inline nilearn/_utils/param_validation.py]: '100' is represented as string. May removing single quotes ?

File: nilearn/_utils/tests/test_validation.py [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+    masks = [None, masker_test_1, masker_test_2, masker_test_3]
+    for mask in masks:
+        masker = check_masker(mask, smoothing_fwhm=smoothing_fwhm, **kwargs)
+        if mask is None:
+            assert_true(isinstance(masker, BaseEstimator))
+        else:
+            assert_true(masker.smoothing_fwhm == smoothing_fwhm_test)
+
+
+def test_feature_screening():
```
Comment (KamalakerDadi) [inline nilearn/_utils/tests/test_validation.py]: test_feature_screening -> test_check_feature_screening

File: nilearn/_utils/tests/test_validation.py [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+from nilearn._utils.testing import assert_warns
+from nilearn._utils.validation import check_masker, check_feature_screening
+from nilearn._utils.param_validation import (MNI152_BRAIN_VOLUME,
+                                             _get_mask_volume)
+
+mni152_brain_mask = (
+    "/usr/share/fsl/data/standard/MNI152_T1_1mm_brain_mask.nii.gz")
+
+
+def test_check_masking():
```
Comment (KamalakerDadi) [inline nilearn/_utils/tests/test_validation.py]: test_check_masking -> test_check_masker

File: nilearn/_utils/validation.py [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+from .param_validation import _adjust_screening_percentile
+from .compat import _basestring
+from ..input_data import NiftiMasker, MultiNiftiMasker
+
+
+def check_masker(mask, target_affine=None, target_shape=None,
+                 smoothing_fwhm=None, standardize=True, t_r=None,
+                 low_pass=None, high_pass=None, mask_strategy='epi',
+                 memory=None, memory_level=1):
+    """Setup a nifti masker.
```
Comment (KamalakerDadi) [inline nilearn/_utils/validation.py]: nifti -> Nifti

File: nilearn/_utils/validation.py [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+from sklearn.base import clone
+from sklearn.feature_selection import (SelectPercentile, f_regression,
+                                       f_classif)
+
+from .param_validation import _adjust_screening_percentile
+from .compat import _basestring
+from ..input_data import NiftiMasker, MultiNiftiMasker
+
+
+def check_masker(mask, target_affine=None, target_shape=None,
```
Comment (KamalakerDadi) [inline nilearn/_utils/validation.py]: >The check_masking is used in two PRs: #1336 and #698

Can you not use `check_embedded_nifti_masker` ?

**Comment by ahoyosid** [comment]
Ping @GaelVaroquaux @bthirion @KamalakerDadi @AlexandreAbraham

File: nilearn/_utils/param_validation.py (hunk 11 above)
Comment (GaelVaroquaux) [inline nilearn/_utils/param_validation.py]: This should be addressed

File: nilearn/_utils/param_validation.py (hunk 16 above)
Comment (GaelVaroquaux) [inline nilearn/_utils/param_validation.py]: +1. It risks being confusing for users.

File: nilearn/decoding/space_net.py [hunk 21]
```diff
@@ -657,10 +604,10 @@ def __init__(self, penalty="graph-net", is_classif=False, loss=None,
                  l1_ratios=.5, alphas=None, n_alphas=10, mask=None,
                  target_affine=None, target_shape=None, low_pass=None,
                  high_pass=None, t_r=None, max_iter=1000, tol=5e-4,
-                 memory=Memory(None), memory_level=1,
-                 standardize=True, verbose=1, n_jobs=1, eps=1e-3,
-                 cv=8, fit_intercept=True, screening_percentile=20.,
-                 debias=False):
+                 memory=None, memory_level=1, standardize=True, verbose=1,
+                 mask_args=None,
```
Comment (GaelVaroquaux) [inline nilearn/decoding/space_net.py]: What is the use of mask_args and where is it used?

Comment (GaelVaroquaux) [inline nilearn/decoding/space_net.py]: Please ignore the above comment.

**Comment by GaelVaroquaux** [comment]
I've added a few comments. Aside from them, the PR looks good, you just need to address the minor comments.

**Comment by GaelVaroquaux** [comment]
Could you please rebase on master: circleCI should be fixed.

**Comment by ahoyosid** [comment]
I already rebased master

**Comment by ahoyosid** [comment]
Ok, this is not passing the circleci. 
In any case, I promise to squash all these noisy commits (when, this passes the tests)

**Comment by bthirion** [comment]
LGTM.

**Comment by GaelVaroquaux** [comment]
Excellent! Merging!
